Sources/NIOCore/AsyncSequences/NIOBackPressuredAsyncSequence.swift

```diff
@@ -0,0 +1,1068 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOConcurrencyHelpers
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// A protocol for the back-pressure strategy of the ``NIOBackPressuredAsyncSequence``.
+///
+/// A back-pressure strategy is invoked when new elements are yielded to the sequence or
+/// when a ``NIOBackPressuredAsyncSequence/AsyncIterator`` requested the next value. The responsibility of the strategy is
+/// to determine whether more elements need to be produced .
+///
+/// If more elements need to be produced, either the ``NIOBackPressuredAsyncSequenceDelegate/produceMore()``
+/// method will be called or a ``NIOBackPressuredAsyncSequence/Source/YieldResult`` will be returned that indicates
+/// to produce more.
+///
+/// - Important: The methods of this protocol are guaranteed to be called serially. Furthermore, the implementation of these
+/// methods **MUST NOT** do any locking or call out to any other Task/Thread.
+public protocol NIOBackPressuredAsyncSequenceStrategy {
+    /// This method is called after new elements were yielded by the producer to the source.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
+    /// - Returns: Returns whether more elements should be produced.
+    mutating func didYield(bufferDepth: Int) -> Bool
+
+    /// This method is called after the `Subscriber` consumed an element.
+    /// More specifically this method is called after `next` was called on an iterator of the ``NIOBackPressuredAsyncSequence``.
+    ///
+    /// - Parameter bufferDepth: The current depth of the internal buffer.
+    /// - Returns: Returns whether the producer should produce more elements.
+    mutating func didConsume(bufferDepth: Int) -> Bool
+}
+
+/// The delegate of ``NIOBackPressuredAsyncSequence``.
+public protocol NIOBackPressuredAsyncSequenceDelegate {
```

Comment:
Same here, `NIOBackPressuredAsyncSequenceDelegate` needs to inherit from Sendable because it is at least passed from the isolation context where it is initialized to the isolation context of the `NIOBackPressuredAsyncSequence`.